File: panic.go
package main

import "fmt"

func main() {
	defer func() {
		if r := recover(); r != nil {
			fmt.Println("panic:", r)
		}
	}()
	for i := 1; i <= 3; i++ {
		text := processText(i)
		fmt.Println(text)
	}
}

var texts = []string{"smile", "tears"}

func processText(id int) string {
	text, err := retrieveText(id)
	if err != nil {
		panic(err)
	}
	codes := []rune(text)
	rotateBack(codes)
	return string(codes)
}

func retrieveText(id int) (string, error) {
	if id >= 1 && id <= len(texts) {
		return texts[id-1], nil
	}
	// return "", fmt.Errorf("404 - Not found: %d", id)
	return "", NotFoundError{Message: fmt.Sprintf("404 - Not found: %d", id)}
}

func rotateBack[T any](vals []T) {
	valsLen := len(vals)
	// if valsLen == 0 {
	// 	return
	// }
	first := vals[0]
	// copy(vals[:valsLen-1], vals[1:])
	for i := 1; i < valsLen; i += 1 {
		vals[i-1] = vals[i]
	}
	vals[valsLen-1] = first
}

type NotFoundError struct {
	Message string
}

func (e NotFoundError) Error() string {
	return e.Message
}

File: panic.zig
const recover = @import("./recover.zig");
const std = @import("std");

pub const panic = recover.panic;

pub fn main() void {
    const allocator = std.heap.page_allocator;
    run(allocator);
    std.debug.print("Still alive.\n", .{});
}

fn run(allocator: std.mem.Allocator) void {
    recover.call(struct {
        pub fn loop(alloc: std.mem.Allocator) void {
            for (1..4) |id| {
                const text = processText(alloc, id);
                defer alloc.free(text);
                std.debug.print("{s}\n", .{text});
            }
        }
    }.loop, .{allocator}) catch |err| {
        std.debug.print("error: {}\n", .{err});
    };
}

fn processText(allocator: std.mem.Allocator, id: usize) []const u8 {
    const text = retrieveText(id) catch @panic("retrieveText failed");
    const codes = stringToCodes(allocator, text) catch @panic("stringToCodes failed");
    defer allocator.free(codes);
    std.mem.reverse(u21, codes);
    return codesToString(allocator, codes) catch @panic("codesToString failed");
}

const texts = [_][]const u8{ "tar", "flow" };
// const texts = [_][]const u8{ "tar", "flow", "🐻‍❄️❤️🦭" };

fn retrieveText(id: usize) error{NotFound}![]const u8 {
    if (id == 0 or id > texts.len) return error.NotFound;
    return texts[id - 1];
}

fn codesToString(allocator: std.mem.Allocator, codes: []const u21) ![]const u8 {
    var buffer = std.ArrayList(u8).init(allocator);
    for (codes) |code| {
        var temp: [4]u8 = undefined;
        const len = try std.unicode.utf8Encode(code, &temp);
        try buffer.appendSlice(temp[0..len]);
    }
    return buffer.toOwnedSlice();
}

fn stringToCodes(allocator: std.mem.Allocator, text: []const u8) ![]u21 {
    var list = std.ArrayList(u21).init(allocator);
    var view = try std.unicode.Utf8View.init(text);
    var it = view.iterator();
    while (it.nextCodepoint()) |code| {
        try list.append(code);
    }
    return list.toOwnedSlice();
}
